README
* description
  weather_catcher is a simple tool to get weather info in the web page
  through parsing the content.
* install and uninstall
  make example-google-hk
  make example-conky
  make install
  make install prefix=/custom/path
  make uninstall
* binary
  /usr/bin/weather_catcher, a wrapper to run the main script weather_catcher.sh
  /usr/bin/weather_catcher_conky_example, just for showing the conky example
* usage
  weather_catcher.sh [-p parser] [-a argument] [-d datatype]
      [-n nightmode] [-f futureday] [-t tempunit] [-l] [-D] [-h] [-v]

  too see complete help message, type "weather_catcher.sh -h"
  too see parser's help message, type "weather_catcher.sh -p google -h"
* examples
  - print all weather info today
    weather_catcher.sh
    weather_catcher.sh -p google -a "weather new york" -d ALL
  - print tomorrow's weather text
    weather_catcher.sh -d WT -f 1
  - print weather font, use night mode
    weather_catcher.sh -d WF -n night
  - use a custom parser
    weather_catcher.sh -p your_parser.sh -a argument
  - print a parser's help message
    weather_catcher.sh -p google -h


Changelog
- 0.2 2013.09.09
  + rename project name from weather-catcher to weathercatch
  + [add] conky screenshots
  + [update] REAME.markdown
- 0.2 2012.11.12
  + [add] parser google
  + [add] parser google-hk
  + [add] parser weather-cn
  + [add] assistant functions
    - convert_url_space()
    - retry_cmd()
    - timeout_cmd()
    - get_section_file()
  + [add] some other files
    - README.markdown
    - Changelog
    - COPYING
    - example_conky_rc
    - configure
    - Makefile.in
	- wrapper.in
	- wrapper_example_conky.in
  + [update] parser baidu build 20121112
  + [update] set google as the default parser
  + [fix] some other simple issues
- 0.1 2012.11.08
  + [add] framework
  + [add] parser baidu


functions.sh
# functions.sh ---
#
# Filename: functions.sh
# Description: assistent functions for weathercatch
# Author: Xu FaSheng
# Created: 2012.11.06
# Last-Updated:
#           By:
#     Update #: 0
# Keywords: assistant function

# This file is a part of weathercatch

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
#
#

# Code:


float_round_off () {
    awk -v float=$1 '
BEGIN {
  CONVFMT="%d";
  i=float+0 "";
  d=float-i;
  if (d<0.5)
    print i;
  else
    print i+1;
}'
}

# temperature, Celsius -> Fahrenheit
temp_c2f () {
    float_round_off `awk "BEGIN {print $1*9/5+32}"`
}

# temperature, Fahrenheit -> Celsius
temp_f2c () {
    float_round_off `awk "BEGIN {print ($1-32)*5/9}"`
}

is_night_now () {
    hour=`date +%H`
    if [[ $hour -ge 18 || `date +%H` -le 6 ]]; then
        echo "t"
    fi
}

# weather font, auto mode
wf_automode () {
    if [ `is_night_now` ]; then
        wf_nightmode "$1"
    else
        wf_daymode "$1"
    fi
}

# weather font, use day mode
wf_daymode () {
    echo "$1" | tr "Olmnopqrstu" "Dbcdefghijk"
}

# weather font, use night mode
wf_nightmode () {
    echo "$1" | tr "Dbcdefghijk" "Olmnopqrstu"
}

set_data_type () {
    name="$1"
    value="$2"
    if [ -z "$data_types_to_write" ]; then
        data_types_to_write="$name=$value"
    else
        data_types_to_write="$data_types_to_write"$'\n'"$name=$value"
    fi
    debug "  $name=$value"
}

write_data_types () {
    # if first time write data, clear the old data
    if [ ! $old_data_is_cleared ]; then
        echo > $weather_data_file
        old_data_is_cleared=t
        debug "old data is cleared"
    fi

    echo "future_day_$1 () {" >> $weather_data_file
    echo "$data_types_to_write" >> $weather_data_file
    echo "}" >> $weather_data_file
    data_types_to_write=
}

do_get_weather_data () {
    # debug "do_get_weather_data: $data_type" ;;TODO
    all_data=`awk "/future_day_$future_day/,/}/" $weather_data_file \
        | sed -e '1d' -e '$d' -e 's/^\s\+//'`
    if [ "$data_type" = "ALL" ]; then
        echo "$all_data"
        return
    fi
    echo "$all_data" | grep -e "^$data_type=" | awk -F= '{print $2}'
}

get_weather_data () {
    # operate hook first
    # don't use "$()" when run get_weather_data_hook() to enable it
    # modify global variables TODO
    debug "data type[before hook]: $data_type"
    get_weather_data_hook
    debug "data type[after hook]: $data_type"
    if [ -n "$hook_result" ]; then
        echo "$hook_result"
        return
    fi

    # get value in data file
    value=`do_get_weather_data`
    debug_lines "value(s):" "$value"

    # result effect
    case "$data_type" in
        WF)                    # weather font, check day or night mode
            value=`general_wf_auto_select_day_or_night "$value"`;;
        LT|HT|CT)               # temperature unit
            # default temp unit in data file should be celsius
            if [ "$temp_unit" = "f" ]; then
                value=`temp_c2f "$value"`
            fi;;
    esac
    echo "$value"
}

# a b c    a 1
# 1 2 3 -> b 2
#          c 3
reverse_table () {
    echo "$1" | awk '
{
  if (max_nf < NF) max_nf=NF
  max_nr=NR
  for(i=1; i<=NF; i++)
    vector[i, NR]=$i
}
END {
  for(i=1; i<=max_nf; i++) {
    for(j=1; j<=max_nr; j++)
      printf("%s ", vector[i, j])
    printf("\n")
  }
}'
}

# a    b           [target]      a  b  c
# 1    [target]    c         ->  1  2  3
#      2           3
clear_w3m_target_block () {
    fixed_table=`echo "$1" | sed "s/^  /$2/"`
    reversed_table=`reverse_table "$fixed_table"`
    cleared_table=`echo "$reversed_table" | awk -v tb=$2 '
{
  for(i=1; i<=NF; i++) {
    if ($i !~ tb)
      printf("%s ", $i)
  }
  printf("\n")
}'`
    reverse_table "$cleared_table"
}

convert_url_space () {
    echo "$1" | sed 's/ /%20/g'
}

retry_cmd () {
    local retry_time=$1
    local cmd="${@:2}"
    for((i=1; i<=$retry_time; i++)); do
        if $cmd; then
            break
        fi
    done
}

timeout_cmd () {
    local timeout=$1
    local cmd="${@:2}"
    # if exist 'timeout' command, use it
    if command -v timeout >/dev/null 2>&1; then
        timeout $timeout $cmd
    else
        $cmd
    fi
}

# output a matched section in a file, support regexp for the start and
# end match word, and include start and end line in output
get_section_file () {
    local file="$1"
    local start="$2"
    local end="$3"
    awk "
/$start/{enable_print=1}
enable_print{print}
/$end/{if (enable_print) exit}
" $file
}

# output a matched section in a multiline sting, support regexp for
# the start and end match word, and include start and end line in
# output
get_section_string () {
    local strings="$1"
    local start="$2"
    local end="$3"
    echo "$strings" | awk "
/$start/{enable_print=1}
enable_print{print}
/$end/{if (enable_print) exit}
"
}

general_wf_auto_select_day_or_night () {
    if [ "$night_mode" = "day" ]; then
        wf_daymode "$1"
    elif [ "$night_mode" = "night" ]; then
        wf_nightmode "$1"
    elif [ "$night_mode" = "auto" ]; then
        wf_automode "$1"
    fi
}

general_weather_text2font_cn () {
    case "$1" in
        *雾*)           echo 'F';;
        *台风*|*飓风*)   echo 'v';;
        *龙卷风*)        echo 'w';;
        *中雪*|*大雪*)   echo 'k';;
        *雪*)           echo 'j';;
        *雷*雨*|*雨*雷*) echo 'i';;
        *雷*)           echo 'f';;
        *中雨*|*大雨*|*暴雨*) echo 'h';;
        *雨*)           echo 'g';;
        *多云转晴*|*晴转多云*) echo 'b';;
        *多云*)         echo 'd';;
        *晴*)           echo 'D';;
        *阴*)           echo 'e';;
        '')echo '';;
        *) echo 'A';;
    esac
}

#
# functions.sh ends here


parser/baidu.sh
# baidu.sh ---
#
# Filename: baidu.sh
# Description: a parser fo weathercatch
# Author: Xu FaSheng
# Created: 2012.11.06
# Keywords: weathercatch parser baidu

# This file is a part of weathercatch

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
#
#

# Code:


parser_name="baidu"
default_arg="天气"
debug "parser_name: $parser_name"

update_weather_data () {
    if [ ! "$arg" ]; then arg="$default_arg"; fi
    URL="www.baidu.com/s?wd=$arg"
    URL=`convert_url_space "$URL"`

    # dump web page
    debug "dumping url: $URL"
    web_content=`retry_cmd 3 timeout_cmd 15s w3m -dump -no-cookie "$URL"`
    if [ -n "$web_content" ]; then
        echo "$web_content">$weather_tmp_file
        parse_data
    fi
}

parse_data () {
    # get valid AQI info section in the page
    aqi_block=`get_section_file $weather_tmp_file "空气质量指数" "空气质量细分数据"`
    debug_lines "AQI block:" "$aqi_block"

    # get AQI value
    # remove first line and empty lines
    aqi_block=`echo "$aqi_block" | sed -e '1d' -e '/^$/d'`
    AQI=`echo "$aqi_block" | head -1 | sed 's/^\([0-9]\+\)$/\1/'`

    # get AQI description
    AQID=`echo "$aqi_block" | head -2 | tail -1`

    # get AQI tips
    AQIT=`echo "$aqi_block" | grep '温馨提示' | sed 's/温馨提示：\(.*\)。/\1/'`

    # get valid weather info section in the page
    weather_block=`get_section_file $weather_tmp_file "一周天气预报" "中国气象局"`
    debug_lines "weather block:" "$weather_block"

    # get location name
    LN=`echo "$weather_block" | head -1 | sed 's/天气预报.*$//'`

    # get other weather data
    weather_block=`echo "$weather_block" | sed -e '1d' -e '$d' | awk 'NF>1'`
    weather_block=`clear_w3m_target_block "$weather_block" "[neverfill]"`
    weather_block=`clear_w3m_target_block "$weather_block" "[u]"`
    debug_lines "weather block[fixed]" "$weather_block"
    weather_block=`reverse_table "$weather_block"`
    i=0
    while read line; do
        debug "future_day_$i"
        set_data_type "LN" $LN

        # temperature
        temp_data=`echo $line | awk '{print $2}'`
        temp_left=`echo $temp_data | awk -F～ '{print $1}' | sed 's/\(-\?[0-9]\+\).*/\1/'`
        temp_right=`echo $temp_data | awk -F～ '{print $2}' | sed 's/\(-\?[0-9]\+\).*/\1/'`
        # there is only low temp in tonight
        if [ -n $temp_right ]; then
            HT=$temp_left
            LT=$temp_right
        else
            HT=
            LT=$temp_left
        fi
        set_data_type "LT" $LT
        set_data_type "HT" $HT

        # weather text
        WT=`echo $line | awk '{print $3}'`
        set_data_type "WT" $WT

        # weather font
        WF=`general_weather_text2font_cn "$WT"`
        set_data_type "WF" $WF

        # wind speed text
        WST=`echo $line | awk '{print $4}'`
        set_data_type "WST" $WST

        # AQI info only exists today
        if [ $i -eq 0 ]; then
            set_data_type "AQI" $AQI
            set_data_type "AQID" $AQID
            set_data_type "AQIT" $AQIT
        fi

        write_data_types $i
        i=$(($i+1))
    done <<< "$weather_block"
}

parser_help () {
    cat <<EOF
Homepage: www.baidu.com
Argument meaning:
    The keyword to search which results will own weather info you
    expect, in fact, it just could be a city name.
    [default: "$default_arg"]
Support data types:
    WF:  weather font output
         [effect by --night-mode]
    WT:  weather text
    LT:  low temperature
         [default unit: celsius]
         [effect by --temp-unit]
    HT:  high temperature
         [default unit: celsius]
         [effect by --temp-unit]
    WST: wind speed text
    LN:  location name
    AQI: AQI(air quality index) value
    AQID:AQI description
    AQIT:AQI tips
Support max future days: 2
Depends:
    w3m v0.5.3+
Compatibility:
    works well on weathercatch_v0.2
    May work on older versions but this is not guaranteed.
Limits:
    Should only work for Chinese friends.
EOF
}

parser_version () {
    echo "build 20130411"
}

debug "parser load success"


#
# baidu.sh ends here


weathercatch.sh
#!/bin/sh

# weathercatch.sh ---
#
# Filename: weathercatch.sh
# Description: a simple tool to get weather info in the web page through parsing the content
# Author: Xu FaSheng
# Created: 2012.11.05
# Keywords: weather web parse
#

# This file is a part of weathercatch

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
#
#

# Code:

help () {
    cat <<EOF
Description:
    weathercatch is a simple tool to get weather info in the web page through parsing the content.

Usage:
    weathercatch.sh [-p parser] [-a argument] [-d datatype] [-n nightmode] [-f futureday] [-t tempunit] [-l] [-D] [-h] [-v]

Options:
    -p, --parser=[PARSER]
        The built-in parser could be:
          google
          baidu
          weather-cn
        And you could use a custom parser, just transmit the
        script file path here.
        [default: $default_parser]
    -a, --argument=[ARGUMENT]
        The argument has different meaning for each parser,
        you should have a look at its help message.
        And the arguments's meaning for built-in parsers could be:
          google         the keyword to search which results will own
                         weather info you expect, in fact, it just
                         could be "weather cityname".
          baidu:         the keyword to search which results will
                         own weather info you expect, in fact, it
                         just could be a city name
          weather-cn:    the complete url which page owns weather
                         info you expect
        [default: different for each parser]
    -d, --data-type=[DATATYPE]
        The built-in data type could be:
          WF:  weather font output
               [effect by --night-mode]
          WT:  weather text
          LT:  low temperature
               [default unit: celsius]
               [effect by --temp-unit]
          HT:  high temperature
               [default unit: celsius]
               [effect by --temp-unit]
          CT:  current temperature
               [default unit: celsius]
               [effect by --temp-unit]
          WD:  wind direction
          WDT: wind direction text
          WS:  wind speed
               [default unit: kmh]
          WST: wind speed text
          HM:  humidity
               [default unit: percent]
          LN:  location name
          ALL: all support data types
        Note: the supported data types is different for each parser,
              you'd better have a look at the parser's help message.
        [default: ALL]
    -n, --night-mode=[NIGHTMODE]
        Will effect some data type's value, such as WF.
        The night mode could be:
          day, night, auto
        [default: auto]
    -f, --future-day=[NUM]
        Get weather info in the future NUM day,
        for example, 0 means today, 1 means tomorrow.
        Note: not all the parser support this options.
        [default: 0]
    -t, --temp-unit=[TEMPUNIT]
        Set the temperature unit, effect for data type HT, LT and CT,
        The temperature unit could be:
          c, celsius
          f, fahrenheit
        [default: c]
    -l, --use-last-data
        Use last weather data if possible.
    -D, --debug
        Use debug mode, for testing.
    -h, --help
        Prints main or parser's help message and exits.
    -v, --version
        Prints main or parser's version information and exits.

Depends:
    w3m, most parsers will use it
    coretuils v7.0+, for the 'timeout' command, not necessary

Examples:
    - print help message
      weathercatch.sh -h
    - print version and built-in parsers
      weathercatch.sh -v
    - print a parser's help message
      weathercatch.sh -p google -h
    - print all weather info today
      weathercatch.sh
      weathercatch.sh -p google -a "weather new york" -d ALL
    - print tomorrow's weather text
      weathercatch.sh -d WT -f 1
    - print weather font, use night mode
      weathercatch.sh -d WF -n night
    - use a custom parser
      weathercatch.sh -p your_parser.sh -a argument

EOF
    exit 1
}

version () {
    echo "version 0.3"
    echo "built-in parsers:"
    for p in ./parser/*.sh; do
        debug $p
        source $p
        printf "  %s: %s\n" "$parser_name" "`parser_version`"
    done
    exit 1
}

debug () {
    if [ $debug ]; then
        echo "[D] " $debug_prefix "$@"
    fi
}

debug_lines () {
    IFS=''
    debug "$1"
    while read line; do
        debug " " "$line"
    done <<< "$2"
}

get_parser_script () {
    builtin_parser="./parser/$1.sh"
    if [ -z "$1" ]; then
        # default parser
        echo "./parser/${default_parser}.sh"
    elif [[ -f "$builtin_parser" && -s "$builtin_parser" ]]; then
        # built-in parser
        echo "$builtin_parser"
    else
        # custom parser
        echo "$1"
    fi
}

## global variables
weather_data_file="/tmp/weathercatch.txt"
weather_tmp_file="/tmp/weathercatch.tmp"
default_parser="google-hk"

## dispatch options
parser=
arg=
data_type=ALL
night_mode=auto
future_day=0
temp_unit=c
use_last_data=
debug=
help=
version=

if ! options=$(getopt -o p:a:d:n:f:t:lDhv -l parser:,argument:,data-type:,night-mode:,future-day:,temp-unit:,use-last-data,debug,help,version -- "$@")
then
    # something went wrong, getopt will put out an error message for us
    exit 1
fi

eval set -- "$options"
while [ $# -gt 0 ]
do
    case $1 in
        -p|--parser)        parser="$2"; shift;;
        -a|--argument)      arg="$2"; shift;;
        -d|--data-type)     data_type="$2"; shift;;
        -n|--night-mode)    night_mode="$2"; shift;;
        -f|--future-day)    future_day="$2"; shift;;
        -t|--temp-unit)     temp_unit="$2"; shift;;
        -l|--use-last-data) use_last_data=t;;
        -D|--debug)         debug=t;;
        -h|--help)          help=t;;
        -v|--version)       version=t;;
        --)                 shift; break ;;
        *)                  break ;;
    esac
    shift
done

# select parser
parser_script=`get_parser_script "$parser"`

# show help message
if [ "$help" ]; then
    if [ "$parser" ]; then
        source "$parser_script"
        parser_help
        exit
    else
        help
    fi
fi
if [ "$version" ]; then
    if [ "$parser" ]; then
        source "$parser_script"
        parser_version
        exit
    else
        version
    fi
fi

## assistant functions
source "./functions.sh"

## variable and function should be redefined in parser
parser_name=
update_weather_data () {
    # the main function for parser, which will get weather info
    # from remote server, then parse it, and put the data into
    # $weather_data_file in a format shows following:
    # future_day_0 () {
    #     WF=a
    #     WT="Sunny"
    #     ...
    # }
    # future_day_1 () {
    #     ...
    # }
    echo "warnning, function need to be redefined"
}
get_weather_data_hook () {
    # if the parser want to operate some special data type, could
    # redefine this function, and change the variable $hook_result,
    # then function 'get_weather_data' will use the output content
    # instead its, and maybe will use the global variable $data_type
    # here.
    hook_result=
}
parser_help () {
    echo "warnning, function need to be redefined"
}
parser_version () {
    echo "warnning, function need to be redefined"
}

## main content
source "$parser_script"

if [[ -f "$weather_data_file" && -s "$weather_data_file" ]]; then
    if [ ! "$use_last_data" ]; then
        update_weather_data
    fi
else
    update_weather_data
fi

get_weather_data


parser/google-hk.sh
# google-hk.sh ---
#
# Filename: google-hk.sh
# Description: a parser fo weathercatch
# Author: Xu FaSheng
# Created: 2012.11.12
# Keywords: weathercatch parser google-hk

# This file is a part of weathercatch

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
#
#

# Code:


# depend on parser google
parent_script=`get_parser_script google`
source "$parent_script"

parser_name="google-hk"
default_arg="天氣"
debug "parser_name: $parser_name"

update_weather_data () {
    if [ -z "$arg" ]; then arg="$default_arg"; fi

    URL="https://www.google.com.hk/search?q=$arg"
    URL=`convert_url_space "$URL"`

    # dump web page
    debug "dumping url: $URL"
    web_content=`retry_cmd 3 timeout_cmd 15s w3m -dump -no-cookie "$URL"`
    if [ -n "$web_content" ]; then
        echo "$web_content">$weather_tmp_file
        parse_data
    fi
}

parse_location () {
    LN=`echo "$weather_block" | head -1 | sed 's/^.*\s\(.*\)天氣$/\1/'`
}

simplify_weather_block () {
    weather_block=`get_section_string "$weather_block" "[0-9]+°C" "濕度："`
    weather_block=`echo "$weather_block" | sed -e 's/^ *• \?//'`
    weather_block=`echo "$weather_block" | sed 's/^\(\S\+\s\)\+//'`

    # delete unicode characters, 0x200e
    weather_block=`echo "$weather_block" | sed 's/‎\(.\)/\1/g'`

    # delete prefix spaces
    weather_block=`echo "$weather_block" | sed 's/^ \+//'`

    debug_lines "weather block[fixed]" "$weather_block"
}

parse_left_weather_block () {
    # get weather tag
    local weather_tag=`echo "$left_weather_block" | sed -n '/\[.*\]/p' | sed 's/\(\[.*\]\).*/\1/'`
    left_weather_block=`echo "$left_weather_block" | sed 's/\[.*\]//'`

    # current weather font
    CWF=`google_weather_tag2wf "$weather_tag"`

    # current temperature
    CT=`echo "$left_weather_block" | sed -n 1p | tr -d '°C'`

    # current weather text
    CWT=`echo "$left_weather_block" | sed -n 2p | sed -e 's/^\s*//' -e 's/\s*$//'`

    # wind direction
    WD=`echo "$left_weather_block" | sed -n 3p | awk -F'，' '{print $1}' | awk -F'：' '{print $2}'`

    # wind speed
    WS=`echo "$left_weather_block" | sed -n 3p | awk -F'，' '{print $2}' | awk -F'：' '{print $2}' | awk '{print $1}'`

    # humidity
    HM=`echo "$left_weather_block" | sed -n '$p' | awk -F'：' '{print $2}' | tr -d '%'`
}

# convert weather tag in google to weather text
google_weather_tag2wt () {
    local tag=`echo "$1" | tr -d '[]'`
    # TODO complete all rules
    case "$tag" in
        fog) echo '有霾';;
        # typhoon|hurricane) echo '';;
        # tornado) echo '';;
        light_s) echo '小雪';;
        sno|snow) echo '雪';;
        tst) echo '雷陣雨';;
        # thunder) echo '';;
        light_r) echo '小雨';;
        rai|rain) echo '雨';;
        par) echo '多雲時晴';;
        partly_) echo '多雲時陰';;
        clo) echo '多雲';;
        cloudy) echo '陰';;
        sun) echo '晴';;
        lig) echo '有雨/雪';;
        '')echo '';;
        *) echo '未知';;
    esac
}

parser_version () {
    echo "build 20130411"
}

debug "parser load success"


#
# google-hk.sh ends here


weather_catcher.sh
#!/bin/sh

# weather_catcher.sh ---
#
# Filename: weather_catcher.sh
# Description: a simple tool to get weather info in the web page through parsing the content
# Author: Xu FaSheng
# Created: 2012.11.05
# Keywords: weather web parse
#

# This file is a part of weather_catcher

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
#
#

# Code:

help () {
    cat <<EOF
Description:
    weather_catcher is a simple tool to get weather info in the web page through parsing the content.

Usage:
    weather_catcher.sh [-p parser] [-a argument] [-d datatype] [-n nightmode] [-f futureday] [-t tempunit] [-l] [-D] [-h] [-v]

Options:
    -p, --parser=[PARSER]
        The built-in parser could be:
          google
          baidu
          weather-cn
        And you could use a custom parser, just transmit the
        script file path here.
        [default: $default_parser]
    -a, --argument=[ARGUMENT]
        The argument has different meaning for each parser,
        you should have a look at its help message.
        And the arguments's meaning for built-in parsers could be:
          google         the keyword to search which results will own
                         weather info you expect, in fact, it just
                         could be "weather cityname".
          baidu:         the keyword to search which results will
                         own weather info you expect, in fact, it
                         just could be a city name
          weather-cn:    the complete url which page owns weather
                         info you expect
        [default: different for each parser]
    -d, --data-type=[DATATYPE]
        The built-in data type could be:
          WF:  weather font output
               [effect by --night-mode]
          WT:  weather text
          LT:  low temperature
               [default unit: celsius]
               [effect by --temp-unit]
          HT:  high temperature
               [default unit: celsius]
               [effect by --temp-unit]
          CT:  current temperature
               [default unit: celsius]
               [effect by --temp-unit]
          WD:  wind direction
          WDT: wind direction text
          WS:  wind speed
               [default unit: kmh]
          WST: wind speed text
          HM:  humidity
               [default unit: percent]
          LN:  location name
          ALL: all support data types
        Note: the supported data types is different for each parser,
              you'd better have a look at the parser's help message.
        [default: ALL]
    -n, --night-mode=[NIGHTMODE]
        Will effect some data type's value, such as WF.
        The night mode could be:
          day, night, auto
        [default: auto]
    -f, --future-day=[NUM]
        Get weather info in the future NUM day,
        for example, 0 means today, 1 means tomorrow.
        Note: not all the parser support this options.
        [default: 0]
    -t, --temp-unit=[TEMPUNIT]
        Set the temperature unit, effect for data type HT, LT and CT,
        The temperature unit could be:
          c, celsius
          f, fahrenheit
        [default: c]
    -l, --use-last-data
        Use last weather data if possible.
    -D, --debug
        Use debug mode, for testing.
    -h, --help
        Prints main or parser's help message and exits.
    -v, --version
        Prints main or parser's version information and exits.

Depends:
    w3m, most parsers will use it
    coretuils v7.0+, for the 'timeout' command, not necessary

Examples:
    - print help message
      weather_catcher.sh -h
    - print version and built-in parsers
      weather_catcher.sh -v
    - print a parser's help message
      weather_catcher.sh -p google -h
    - print all weather info today
      weather_catcher.sh
      weather_catcher.sh -p google -a "weather new york" -d ALL
    - print tomorrow's weather text
      weather_catcher.sh -d WT -f 1
    - print weather font, use night mode
      weather_catcher.sh -d WF -n night
    - use a custom parser
      weather_catcher.sh -p your_parser.sh -a argument

EOF
    exit 1
}

version () {
    echo "version 0.2"
    echo "built-in parsers:"
    for p in ./parser/*.sh; do
        debug $p
        source $p
        printf "  %s: %s\n" "$parser_name" "`parser_version`"
    done
    exit 1
}

debug () {
    if [ $debug ]; then
        echo "[D] " $debug_prefix "$@"
    fi
}

debug_lines () {
    IFS=''
    debug "$1"
    while read line; do
        debug " " "$line"
    done <<< "$2"
}

get_parser_script () {
    builtin_parser="./parser/$1.sh"
    if [ -z "$1" ]; then
        # default parser
        echo "./parser/${default_parser}.sh"
    elif [[ -f "$builtin_parser" && -s "$builtin_parser" ]]; then
        # built-in parser
        echo "$builtin_parser"
    else
        # custom parser
        echo "$1"
    fi
}

## global variables
weather_data_file="/tmp/weather_catcher.txt"
weather_tmp_file="/tmp/weather_catcher.tmp"
default_parser="google-hk"

## dispatch options
parser=
arg=
data_type=ALL
night_mode=auto
future_day=0
temp_unit=c
use_last_data=
debug=
help=
version=

if ! options=$(getopt -o p:a:d:n:f:t:lDhv -l parser:,argument:,data-type:,night-mode:,future-day:,temp-unit:,use-last-data,debug,help,version -- "$@")
then
    # something went wrong, getopt will put out an error message for us
    exit 1
fi

eval set -- "$options"
while [ $# -gt 0 ]
do
    case $1 in
        -p|--parser)        parser="$2"; shift;;
        -a|--argument)      arg="$2"; shift;;
        -d|--data-type)     data_type="$2"; shift;;
        -n|--night-mode)    night_mode="$2"; shift;;
        -f|--future-day)    future_day="$2"; shift;;
        -t|--temp-unit)     temp_unit="$2"; shift;;
        -l|--use-last-data) use_last_data=t;;
        -D|--debug)         debug=t;;
        -h|--help)          help=t;;
        -v|--version)       version=t;;
        --)                 shift; break ;;
        *)                  break ;;
    esac
    shift
done

# select parser
parser_script=`get_parser_script "$parser"`

# show help message
if [ "$help" ]; then
    if [ "$parser" ]; then
        source "$parser_script"
        parser_help
        exit
    else
        help
    fi
fi
if [ "$version" ]; then
    if [ "$parser" ]; then
        source "$parser_script"
        parser_version
        exit
    else
        version
    fi
fi

## assistant functions
source "./functions.sh"

## variable and function should be redefined in parser
parser_name=
update_weather_data () {
    # the main function for parser, which will get weather info
    # from remote server, then parse it, and put the data into
    # $weather_data_file in a format shows following:
    # future_day_0 () {
    #     WF=a
    #     WT="Sunny"
    #     ...
    # }
    # future_day_1 () {
    #     ...
    # }
    echo "warnning, function need to be redefined"
}
get_weather_data_hook () {
    # if the parser want to operate some special data type, could
    # redefine this function, and change the variable $hook_result,
    # then function 'get_weather_data' will use the output content
    # instead its, and maybe will use the global variable $data_type
    # here.
    hook_result=
}
parser_help () {
    echo "warnning, function need to be redefined"
}
parser_version () {
    echo "warnning, function need to be redefined"
}

## main content
source "$parser_script"

if [[ -f "$weather_data_file" && -s "$weather_data_file" ]]; then
    if [ ! "$use_last_data" ]; then
        update_weather_data
    fi
else
    update_weather_data
fi

get_weather_data
